Kenneth l/otp 25 blog highlights (#1)
KennethL/erlang-org: _posts/2022-05-18-My-OTP-25-highlights.md

---
layout: post
title: Erlang/OTP 25 Highlights
tags: erlang otp 25 release
author: Kenneth Lundin
---

OTP 25 is finally here. This post will introduce the new features that I am most excited about.

Erlang/OTP 25 includes contributions from ??+ external contributors totalling
??+ commits, ??+ PRs.

You can download the readme describing all the changes here: [Erlang/OTP 25 Readme].
Or, as always, look at the release notes of the application you are interested in.
For instance here: [Erlang/OTP 25 - Erts Release Notes - Version 13.0].

This years highlights are:
* [New functions in the `maps`and `lists` modules](#new-functions-in-the-mapsand-lists-modules)
* [Selectable features and the new `maybe_expr` feature](#selectable-features-and-the-new-maybe_expr-feature)
* Compiler News
* Crypto and OpenSSL 3.0
* Dialyzer
* [Improvements of the JIT](#improvements-of-the-jit)
* ETS-tables with adaptive support for write concurrency
* Relocatable installation directory for Erlang
* New option `short` to the functions `erlang:float_to_list/2` and `erlang:float_to_binary/2` 
* The new module `peer` supersedes the slave module
* global will now by default prevent overlapping partitions

> * gen_server, gen_statem and gen_event has got a new format_status/1 callback.

Reviewer: ```suggestion
* `global` will now by default prevent overlapping partitions
* `gen_server`, `gen_statem`, and `gen_event` have a new `format_status/1` callback.
```

* The `timer` module has been modernized and made more efficient
* CA-certificates can be fetched from the OS standard place 

# New functions in the `maps`and `lists` modules

> Triggered by suggestions from the users we have introduced new functions in the `maps` and `lists` modules in `stdlib.

Reviewer: ```suggestion
Triggered by suggestions from the users we have introduced new functions in the `maps` and `lists` modules in `stdlib`.
```

## `maps:groups_from_list/2,3`

For short we can say that this function take a list of elements and group them. The result is a map `#{Group1 => [Group1Elements], GroupN => [GroupNElements]}`.

Let us look at some examples from the shell:

```erlang
> maps:groups_from_list(fun(X) -> X rem 2 end, [1,2,3]).
#{0 => [2], 1 => [1, 3]}
```

> The provided fun calculates X rem 2 for every element X in the input list and then group the elements in a map with the result of X rem 2 as key and the corresponding elements as a list for that key.

Reviewer: ```suggestion
The provided fun calculates `X rem 2` for every element `X` in the input list and then groups the elements in a map with the result of `X rem 2` as key and the corresponding elements as a list value for that key.
```

```erlang
> maps:groups_from_list(fun erlang:length/1, ["ant", "buffalo", "cat", "dingo"]).
#{3 => ["ant", "cat"], 5 => ["dingo"], 7 => ["buffalo"]}
```

In the example above the strings in the input list are grouped into a map based on their length.

## `maps:groups_from_list/3`

> There is also a variant of groups_from_list with an additional fun by which the values can be converted or changed before they are put into their groups.

Reviewer: ```suggestion
There is also a variant of `groups_from_list` with an additional fun by which the values can be converted or changed before they are put into their groups.
```

```erlang
> maps:groups_from_list(fun(X) -> X rem 2 end, fun(X) -> X*X end, [1,2,3]).
#{0 => [4], 1 => [1, 9]}
```

> In the example above the elements X in the list are grouped according the X rem 2 calculation but the values stored in the groups are the elements multiplied by them selves (X*X).  

Reviewer: ```suggestion
In the example above the elements `X` in the list are grouped according the `X rem 2` calculation but the values stored in the groups are the elements multiplied by themselves (`X*X`).  
```

```erlang
> maps:groups_from_list(fun erlang:length/1, fun lists:reverse/1, ["ant", "buffalo", "cat", "dingo"]).
#{3 => ["tna","tac"],5 => ["ognid"],7 => ["olaffub"]}
```

In the example above the strings from the input list are grouped according to their length and they are reversed before they are stored in the groups.

## `lists:enumerate/1,2`

Takes a list of elements and return a new list where each element has been associated with its position in the original list. Returns a new list with tuples of the form `{I, H}` where `I` is the position of `H` in the original list. The enumeration starts with 1 and increases by 1 in each step.

> That is, enumerate/1 behaves as if it had been defined as follows:

Reviewer: ```suggestion
That is, `enumerate/1` behaves as if it had been defined as follows:
```

Something seems to be missing here. How is it defined?

Example:
```erlang
> lists:enumerate([a,b,c]).
[{1,a},{2,b},{3,c}]
```

> There is also a enumerate/2 function which can be used to set the initial number to something else than 1. See example below:

Reviewer: ```suggestion
There is also an `enumerate/2` function which can be used to set the initial number to something else than 1. See example below:
```

[{10,a},{11,b},{12,c}]
```

## `lists:uniq/1,2`

Removes duplicates from a list while preserving the order of the elements. The first occurrence of each element is kept. 
We already have `lists:usort` which also removes duplicates but return a sorted list.

Examples:
```erlang
> lists:uniq(fun([3,3,1,2,1,2,3]).
[3,1,2]
> lists:uniq([a, a, 1, b, 2, a, 3]).
[a, 1, b, 2, 3]
```
`lists:uniq/2` allows the user to specify with a fun how to determine that 2 elements in the list are equal. In the example below the provided fun is just testing the first element of the 2 tuples for equality.

Examples:
```erlang
> lists:uniq(fun({X, _}) -> X end, [{b, 2}, {a, 1}, {c, 3}, {a, 2}]).
[{b, 2}, {a, 1}, {c, 3}]
```
# Selectable features and the new `maybe_expr` feature

Selectable features is a new mechanism and concept where a new potentially incompatible feature (language or runtime), can be introduced and tested without causing troubles for those that don't use it.

When it comes to language features the intention is that they can be activated per module with no impact on modules where they are not activated.

Let's use the new `maybe_expr` feature as an example.

> In module my_experiment the feature is activated and used like this:

Reviewer: ```suggestion
In module `my_experiment` the feature is activated and used like this:
```

```erlang
-module(my_experiment).
-export([foo/1]).

> %% enable the feature maybe_expr in this module only
> %% makes maybe a keyword which might be incompatible 
> %% in modules using maybe as a function name or an atom  

Reviewer: ```suggestion
%% Enable the feature maybe_expr in this module only.
%% Makes maybe a keyword which might be incompatible 
%% in modules using maybe as a function name or an atom.
```

-feature(maybe_expr,enable) 
foo() ->
  maybe
    {ok, X} ?= f(Foo),
    [H|T] ?= g([1,2,3]),
    ...
  else
    {error, Y} ->
        {ok, "default"};
    {ok, _Term} ->
        {error, "unexpected wrapper"}

> end

Reviewer 2: ```suggestion
  end.
```

```

The compiler will note that the feature `maybe_expr` is enabled and will handle the maybe construct correctly. In the generated `.beam` file it will also be noted that
the module has enabled the feature.

When starting an Erlang node the specific feature (or all) must be enabled otherwise the `.beam` file with the feature will not be allowed for loading.

```
erl -enable-feature maybe_expr
```

Or 

```
erl -enable-feature all
```

## The new `maybe_expr` feature EEP-49

> The EEP-49 "Value-Based Error Handling Mechanisms", was suggested by Fred Hebèrt already 2018 and now finally
> it has been implemented as the first feature within the new feature concept.

Reviewer: ```suggestion
The EEP-49 "Value-Based Error Handling Mechanisms", was suggested by Fred Hebèrt already in 2018 and
now it has finally been implemented as the first feature within the new feature concept.
```

for variables (`OpaqueData` renamed to `B`), and half of the code just
transparently returns the exact values each function was given.

By comparison, the same code could be written as follows with the new
construct:

```erlang
commit_write(OpaqueData) ->
    maybe
        ok ?= disk_log:sync(OpaqueData#backup.file_desc),
        ok ?= disk_log:close(OpaqueData#backup.file_desc),
        ok ?= file:rename(OpaqueData#backup.tmp_file, OpaqueData#backup.file),
        {ok, OpaqueData#backup.file}
    end.
```

Or, to protect against `disk_log` calls returning something else than `ok |
{error, Reason}`, the following form could be used:

```erlang
commit_write(OpaqueData) ->
    maybe
        ok ?= disk_log:sync(OpaqueData#backup.file_desc),
        ok ?= disk_log:close(OpaqueData#backup.file_desc),
        ok ?= file:rename(OpaqueData#backup.tmp_file, OpaqueData#backup.file),
        {ok, OpaqueData#backup.file}
    else
        {error, Reason} -> {error, Reason}
    end.
```

The semantics of these calls are identical, except that it is now
much easier to focus on the flow of individual operations and either
success or error paths.

> # Compiler news

Reviewer 2: The binary syntax things are described in detail later in the post. Not sure the -nifs attribute qualifies for a highlight.

Author: Agree on this too, removed

Author: agree, removed

* Add compile attribute `-nifs()` to empower compiler and loader with information     about   which functions may be overridden as NIFs by `erlang:load_nif/2`.
* Improved and more detailed error messages when binary construction with the binary syntax fails. This applies both for error messages in the shell and for `erl_error:format_exception/3,4`.

# Crypto and OpenSSL 3.0
Text to be included.....

# Dialyzer

> Optimize operations in the erl_types module. Parallelize the Dialyzer pass remote.
> Added the missing_return and extra_return options to raise warnings when specifications differ from inferred types. These are similar to, but not quite as verbose as overspecs and underspecs.
> Dialyzer now better understands the types for min/2, max/2, and erlang:raise/3. Because of that, Dialyzer can potentially generate new warnings. In particular, functions that use erlang:raise/3 could now need a spec with a no_return() return type to avoid an unwanted warning.

Reviewer: ```suggestion
Dialyzer now supports the `missing_return` and `extra_return` options to raise warnings when specifications differ from inferred types. These are similar to, but not quite as verbose, as `overspecs` and `underspecs`.

Dialyzer now better understands the types for `min/2`, `max/2`, and `erlang:raise/3`. Because of that, Dialyzer can potentially generate new warnings. In particular, functions that use `erlang:raise/3` could now need a spec with a `no_return()` return type to avoid an unwanted warning.
```

at it in the source code to see if you can figure out why so much time is spent there.

After `eq` we see the function `erl_types:t_has_var/1` where we spend almost
5% of the entire execution in. A while further down you can see `copy_struct_x`
which is the function used to copy terms. If we expand it to view the parents
we find that it is mostly `ets:lookup_element/3` that contributes to this time
via the Erlang function `dialyzer_plt:ets_table_lookup/2`.

### perf tips and tricks

You can do a lot of neat things with `perf`. Below is a list of some of the options
we have found useful:

* `perf report --no-children`
    Do not include the accumulation of all children in a call.
* `perf report  --call-graph callee`
    Show the callee rather than the caller when expanding a function call.
* `perf archive`
    Create an archive with all the artifacts needed to inspect the data
    on another host. In early version of perf this command does not work,
    instead you can use [this bash script](https://github.com/torvalds/linux/blob/master/tools/perf/perf-archive.sh).
* `perf report` gives "failed to process sample" and/or "failed to process type: 68"

>     This probably means that you are running a bugge version of perf. We have

Reviewer 2: ```suggestion
    This probably means that you are running a buggy version of perf. We have
```

    seen this when running Ubuntu 18.04 with kernel version 4. If you update
    to Ubuntu 20.04 or use Ubuntu 18.04 with kernel version 5 the problem

>     should go away.

Reviewer 2: ```suggestion
    should go away.

```

# Improved error information for failing binary construction

Erlang/OTP 24 introduced [improved BIF error information][ext_bif_info] to provide
more information when a call to a BIF failed.

[ext_bif_info]: https://www.erlang.org/blog/my-otp-24-highlights/#eep-54-improved-bif-error-information

In Erlang/OTP 25, improved error infomation is also given when the
creation of a binary using the [bit syntax][bit_syntax] fails.

Consider this function:

> ```

Reviewer 2: ```suggestion
```erlang
```

     in function  t:rec_add/1 (t.erl, line 8)
```

The new `badrecord` exceptions will show up for code that has been compiled
with Erlang/OTP 25.

# Relocatable installation directory

Previously shell scripts (e.g., `erl` and `start`) and the `RELEASES` file
for an Erlang installation depended on a hard coded absolute path to the
installation's root directory. This made it cumbersome to move an
installation to a different directory which can be problematic for platforms
such as Android (#2879) where the
installation directory is unknown at compile time. This is fixed by:

* Changing the shell scripts so that they can dynamically find the
  `ROOTDIR`. The dynamically found `ROOTDIR` is selected if it differs
  from the hard-coded `ROOTDIR` and seems to point to a valid Erlang
  installation. The `dyn_erl` program has been changed so that it can
  return its absolute canonicalized path when given the --realpath
  argument (dyn_erl gets its absolute canonicalized path from the
  realpath POSIX function). The dyn_erl's --realpath
  functionality is used by the scripts to get the root dir dynamically.

* Changing the release_handler module that reads and writes to the

>   RELEASES file so that it prepends code:root_dir() whenever it

Reviewer: ```suggestion
  `RELEASES` file so that it prepends `code:root_dir()` whenever it
```

  encounters relative paths. This is necessary since the current
  working directory can be changed so it is something different than
  `code:root_dir()`.

# ETS-tables with adaptive support for write concurrency

> It has since long been possible to optimize an ets-table for write concurrency doing like this:

Reviewer: ```suggestion
It has since long been possible to optimize an ETS table for write concurrency doing like this:
```

Benchmark results comparing this option with the other ETS optimization options are available here: benchmarks.

# New option `short` for `erlang:float_to_list/2` and `erlang:float_to_binary/2` 

A new option called `short` has been added to the functions `erlang:float_to_list/2` and `erlang:float_to_binary/2`. This option creates the shortest correctly rounded string representation of the given float that can be converted back to the same float again.

If option `short` is specified, the float is formatted
with the smallest number of digits that still guarantees that
```erlang
F =:= list_to_float(float_to_list(F, [short]))
``` 
When the float is inside the range (-2⁵³, 2⁵³), the notation
that yields the smallest number of characters is used (scientific
notation or normal decimal notation). Floats outside the range
(-2⁵³, 2⁵³) are always formatted using scientific notation to avoid confusing 
results when doing arithmetic operations.

> The implementation is contributed by Thomas Depierre and uses the Ryu algorithm.

Reviewer 2: ```suggestion
The implementation is contributed by Thomas Depierre and uses the Ryū algorithm.
```

Peer node terminal input/output is relayed through the origin. If a standard I/O alternative connection is requested, console output also goes via the origin, allowing debugging of node startup and boot script execution (see -init_debug). File I/O is not redirected, contrary to slave(3) behaviour.

The peer node can start on the same or a different host (via ssh) or in a separate container (for example Docker). When the peer starts on the same host as the origin, it inherits the current directory and environment variables from the origin.

## Note

This module is designed to facilitate multi-node testing with Common Test. Use the ?CT_PEER() macro to start a linked peer node according to Common Test conventions: crash dumps written to specific location, node name prefixed with module name, calling function, and origin OS process ID). Use `random_name/1` to create sufficiently unique node names if you need more control.

A peer node started without alternative connection behaves similarly to `slave(3)`. When an alternative connection is requested, the behavior is similar to 

> `test_server:start_node(Name, peer, Args)`. 

Reviewer 2: Do we really want to mention `test_server:start_node`? It is an internal API that should(?) only have been used in our own tests.

# global will now by default prevent overlapping partitions

> # gen_xxx modules has got a new format_status/1 callback.

Reviewer 2: I would drop this from the highlights

Author: Agree , removed

---
layout: post
title: Erlang/OTP 25 Highlights
tags: erlang otp 25 release
author: Kenneth Lundin
---

OTP 25 is finally here. This post will introduce the new features that I am most excited about.

Erlang/OTP 25 includes contributions from ??+ external contributors totalling
??+ commits, ??+ PRs.

You can download the readme describing all the changes here: [Erlang/OTP 25 Readme].
Or, as always, look at the release notes of the application you are interested in.
For instance here: [Erlang/OTP 25 - Erts Release Notes - Version 13.0].

This years highlights are:

> * [New functions in the `maps`and `lists` modules](#new-functions-in-the-mapsand-lists-modules)

Reviewer 2: ```suggestion
* [New functions in the `maps`and `lists` modules](#new-functions-in-the-maps-and-lists-modules)
```

* [Selectable features and the new `maybe_expr` feature](#selectable-features-and-the-new-maybe_expr-feature)
* Compiler News
* Crypto and OpenSSL 3.0
* Dialyzer
* [Improvements of the JIT](#improvements-of-the-jit)
* ETS-tables with adaptive support for write concurrency
* Relocatable installation directory for Erlang
* New option `short` to the functions `erlang:float_to_list/2` and `erlang:float_to_binary/2` 
* The new module `peer` supersedes the slave module
* global will now by default prevent overlapping partitions
* gen_server, gen_statem and gen_event has got a new format_status/1 callback.
* The `timer` module has been modernized and made more efficient
* CA-certificates can be fetched from the OS standard place 

> # New functions in the `maps`and `lists` modules

Reviewer 2: ```suggestion
# New functions in the `maps` and `lists` modules
```

> Triggered by suggestions from the users we have introduced new functions in the `maps` and `lists` modules in `stdlib.

Reviewer 2: ```suggestion
Triggered by suggestions from the users we have introduced new functions in the [`maps`](/doc/man/maps.html) and [`lists`](/doc/man/lists.html) modules in `stdlib`.
```

## `maps:groups_from_list/2,3`

For short we can say that this function take a list of elements and group them. The result is a map `#{Group1 => [Group1Elements], GroupN => [GroupNElements]}`.

Let us look at some examples from the shell:

```erlang
> maps:groups_from_list(fun(X) -> X rem 2 end, [1,2,3]).
#{0 => [2], 1 => [1, 3]}
```
The provided fun calculates X rem 2 for every element X in the input list and then group the elements in a map with the result of X rem 2 as key and the corresponding elements as a list for that key.

```erlang
> maps:groups_from_list(fun erlang:length/1, ["ant", "buffalo", "cat", "dingo"]).
#{3 => ["ant", "cat"], 5 => ["dingo"], 7 => ["buffalo"]}
```

In the example above the strings in the input list are grouped into a map based on their length.

> ## `maps:groups_from_list/3`

Reviewer 2: ```suggestion
```

There is already a title for maps:group_from_list/2,3. So this title is redundant.

> There is also a variant of groups_from_list with an additional fun by which the values can be converted or changed before they are put into their groups.

Reviewer 3: What's the difference between 'converted' and 'changed' ?

Author: Unnecessary to have both. Removed changed.

```erlang
> maps:groups_from_list(fun(X) -> X rem 2 end, fun(X) -> X*X end, [1,2,3]).
#{0 => [4], 1 => [1, 9]}
```
In the example above the elements X in the list are grouped according the X rem 2 calculation but the values stored in the groups are the elements multiplied by them selves (X*X).  

```erlang
> maps:groups_from_list(fun erlang:length/1, fun lists:reverse/1, ["ant", "buffalo", "cat", "dingo"]).
#{3 => ["tna","tac"],5 => ["ognid"],7 => ["olaffub"]}
```

> In the example above the strings from the input list are grouped according to their length and they are reversed before they are stored in the groups.

Reviewer 2: ```suggestion
In the example above the strings from the input list are grouped according to their length and they are reversed before they are stored in the groups.

For more details see the [`maps:groups_from_list/2`](/doc/man/maps.html#groups_from_list-2) documentation.
```

## `lists:enumerate/1,2`

> Takes a list of elements and return a new list where each element has been associated with its position in the original list. Returns a new list with tuples of the form `{I, H}` where `I` is the position of `H` in the original list. The enumeration starts with 1 and increases by 1 in each step.

Reviewer 3: "Takes a list of elements and return ..."  'Takes' is 3rd person singular but 'return' is not although same subject.  It should it be 'return**s**'

That is, enumerate/1 behaves as if it had been defined as follows:

Example:
```erlang
> lists:enumerate([a,b,c]).
[{1,a},{2,b},{3,c}]
```
There is also a enumerate/2 function which can be used to set the initial number to something else than 1. See example below:
```erlang
> lists:enumerate(10, [a,b,c]).
[{10,a},{11,b},{12,c}]
```

> 

Reviewer 2: ```suggestion
For more details see the [`lists:enumerate/1`](/doc/man/lists.html#enumerate-1) documentation.
```

## `lists:uniq/1,2`

Removes duplicates from a list while preserving the order of the elements. The first occurrence of each element is kept. 

> We already have `lists:usort` which also removes duplicates but return a sorted list.

Reviewer 3: return**s**

> We already have `lists:usort` which also removes duplicates but return a sorted list.

Reviewer 3: returns

Examples:
```erlang
> lists:uniq(fun([3,3,1,2,1,2,3]).
[3,1,2]
> lists:uniq([a, a, 1, b, 2, a, 3]).
[a, 1, b, 2, 3]
```
`lists:uniq/2` allows the user to specify with a fun how to determine that 2 elements in the list are equal. In the example below the provided fun is just testing the first element of the 2 tuples for equality.

Examples:
```erlang
> lists:uniq(fun({X, _}) -> X end, [{b, 2}, {a, 1}, {c, 3}, {a, 2}]).
[{b, 2}, {a, 1}, {c, 3}]
```

> # Selectable features and the new `maybe_expr` feature

Reviewer 2: ```suggestion

For more details see the [`lists:uniq/1`](/doc/man/lists.html#uniq-1) documentation.

# Selectable features and the new `maybe_expr` feature
```

When it comes to language features the intention is that they can be activated per module with no impact on modules where they are not activated.

Let's use the new `maybe_expr` feature as an example.

In module my_experiment the feature is activated and used like this:

```erlang
-module(my_experiment).
-export([foo/1]).

%% enable the feature maybe_expr in this module only
%% makes maybe a keyword which might be incompatible 
%% in modules using maybe as a function name or an atom  
-feature(maybe_expr,enable) 
foo() ->
  maybe
    {ok, X} ?= f(Foo),
    [H|T] ?= g([1,2,3]),
    ...
  else
    {error, Y} ->
        {ok, "default"};
    {ok, _Term} ->
        {error, "unexpected wrapper"}
end
```

The compiler will note that the feature `maybe_expr` is enabled and will handle the maybe construct correctly. In the generated `.beam` file it will also be noted that
the module has enabled the feature.

When starting an Erlang node the specific feature (or all) must be enabled otherwise the `.beam` file with the feature will not be allowed for loading.

```
erl -enable-feature maybe_expr
```

Or 

```
erl -enable-feature all
```

> 

Reviewer 2: ```suggestion

For more details see the [feature section](/doc/reference_manual/feature.html) in the Erlang Reference Manual.

```

environment, and the expression is equivalent to a successful `Pattern = Expr`
call. If the value does not match, the `maybe ... end` expression returns the
failed expression directly.

A special case exists in which we extend `maybe ... end` into the following form:

```
maybe
    Exprs | MatchOrReturnExprs
else
    Pattern -> Exprs;
    ...
    Pattern -> Exprs
end
```

This form exists to capture non-matching expressions in a `MatchOrReturnExprs`
to handle failed matches rather than returning their value. In such a case, an
unhandled failed match will raise an `else_clause` error, otherwise identical to
a `case_clause` error.

This extended form is useful to properly identify and handle successful and
unsuccessful matches within the same construct without risking to confuse
happy and unhappy paths.

Given the structure described here, the final expression may look like:

```erlang
maybe
    Foo = bar(),            % normal exprs still allowed
    {ok, X} ?= f(Foo),
    [H|T] ?= g([1,2,3]),
    ...
else
    {error, Y} ->
        {ok, "default"};
    {ok, _Term} ->
        {error, "unexpected wrapper"}
end
```

> 

Reviewer 2: ```suggestion

For more details see the [maybe section](/doc/reference_manual/expressions.html#maybe) in the Erlang Reference Manual.

```

They will then be cached in decoded form by use of `persistent_term` which makes them available in an efficient way for the `ssl` and `httpc` modules. The intention with this is to make it unnecessary to depend on for example `certifi` in many packages.

On Windows and MacOSx the certificate store is not an ordinary file so the information is fetched via an API using a NIF (Windows) or with an external program (MacOSx). 

Example with `ssl`
```erlang
public_key:cacerts_load(), % a noop if certs are already loaded
%% makes the certificates available without copying
CaCerts = public_key:cacerts_get(), 
% use the certificates when establishing a connection
{ok,Socket} = ssl:connect("erlang.org",80,[{cacerts,CaCerts}, {verify,verify_peer}]), 
...
```
We also plan to update the http client (`httpc`) to use this soon.

# A new fast Pseudo Random Generator

A new custom designed Pseudo Random Generator `rand:mwc59`
has been implemented.  It is probably the fastest possible
generator with good quality that can be written in Erlang.
To do this it barely avoids bignums, allocating heap data,
and uses only a minimal number of fast operations.

> 

Reviewer 4: Suggested addition:

Under the "right" circumstances: A number that takes 60 ns to generate
with the default generator can be generated in 4 ns with `mwc59`.

Author: Added this

Author: Added

0.9.8c and later, although only 1.1.1 is still maintained by OpenSSL.

OpenSSL has released its version 3.0 series, which is their future
platform totally re-built with a new API. The APIs of previous
versions (1.1.1 and older) are partly deprecated, although still
available in 3.0. The support of 1.1.1 will also end in a future.

Since it is vital to get security patches in the cryptolib, and in a
future only the 3.0 API might be available, OTP/crypto now from
OTP-25.0 interfaces OpenSSL 3.0 using the new 3.0 API. A few functions
from old APIs are still used, but they will be replaced as soon as
possible.

You as a user will hopefully not notice any difference: if you have
OpenSSL 1.1.1 (or older - not recommended) and build OTP, that one will
be used as previously. If you have any OpenSSL 3.0 version installed,
that one will be used without need of doing anything special except
for normal handling of dynamic loading paths in the OS.

Although OTP/crypto's interface to the new API is heavily tested on many
architectures for a long time, we do not yet recommend it for usage in
critical applications. In such ones it is still safer to build crypto with

> the latest 1.1.1 release.

Reviewer 3: Although I've written this, I now think we should remove it.  This is a blog post - not a legal document.

That warning is present in the release notes and you also get it when running configure with a 3.0 lib.  That is thought as a kind of disclaimer if anyone complains that something doesn't work.

Author: Agree, removed
